<a href="https://colab.research.google.com/github/VrushaliLad799/707-lecture-master/blob/main/Assignment_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Assignment 4**

Did you follow the rules? If not, your assignment will be returned ungraded.

Notebook organization:
Two major sections, one for each dataset. Each section contains:
Data generation code (5 points)
Data preprocessing as necessary, including scaling, encoding (5 points)
Proper model execution with 5-fold cross validation on each model (10 points)

Did you achieve at least a .02 difference in performance on both datasets?

In [76]:
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression, SGDClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

# M1 = LogisticRegression,
# M2 = DecisionTreeClassifier

m1 = LogisticRegression()
m2 = DecisionTreeClassifier()

# Dataset 1 - Favors Logistic Regression
X1, y1 = make_classification(n_samples=1000, n_features=20, n_informative=10, n_redundant=3, n_clusters_per_class=1, random_state=10)

# Scale dataset 1
scaler = StandardScaler()
X1 = scaler.fit_transform(X1)

#Cross val score with f1 as scoring
m1_scores_A = cross_val_score(m1, X1, y1, cv=5, scoring='f1')
m2_scores_A = cross_val_score(m2, X1, y1, cv=5, scoring='f1')

print("Dataset 1:")
print("Logistic Regression F1: {:.2f}".format(m1_scores_A.mean()))
print("Decision Tree Classifier F1: {:.2f}".format(m2_scores_A.mean()))

Dataset 1:
Logistic Regression F1: 0.93
Decision Tree Classifier F1: 0.87


In [77]:
# Dataset 2 - Decision Tree Classifier
X2, y2 = make_classification(n_samples=10000, n_features=100, n_informative=4, n_redundant=50, n_clusters_per_class=4, random_state=10)


# Scale dataset 2
scaler = StandardScaler()
X2 = scaler.fit_transform(X2)

#Cross val score with f1 as scoring
m1_scores_B = cross_val_score(m1, X2, y2, cv=5, scoring='f1')
m2_scores_B = cross_val_score(m2, X2, y2, cv=5, scoring='f1')

print("\nDataset 2:")
print("Logistic Regression F1: {:.2f}".format(m1_scores_B.mean()))
print("Decision Tree Classifier F1: {:.2f}".format(m2_scores_B.mean()))


Dataset 2:
Logistic Regression F1: 0.67
Decision Tree Classifier F1: 0.87


In [79]:
# Saving datasets A and B as csv files
import pandas as pd

# For dataset A
data_A = pd.DataFrame(X1)
data_A['target'] = y1
data_A.to_csv('data_A.csv', index=False)

# For dataset B
data_B = pd.DataFrame(X2)
data_B['target'] = y2
data_B.to_csv('data_B.csv', index=False)

***Dataset 1:***

Logistic Regression F1: **0.93**

Decision Tree Classifier F1: **0.87**

Difference: **0.06**

***Dataset 2:***

Logistic Regression F1: **0.67**

Decision Tree Classifier F1: **0.87**

Difference: **0.20**

From above we can see Logistic Regression performs best on Dataset 1 whereas Decision Tree Classifier performs poor.

The decision Tree Classifier performs best on Dataset 2 whereas Logistic Regression performs poor.

The minimal discrepancy is in Dataset 1 which is **0.06**

***Findings:***


Dataset 1 favors Logistic Regression (M1) because it has 10 informative features and only 3 redundant ones. Logistic Regression works best with a moderate number of informative features and fewer irrelevant ones.

Dataset 2 favors Decision Tree (M2) because it has only 4 informative features out of 100 total features, and a significant number (50) of redundant features. Decision Trees can handle large numbers of irrelevant features and complex interactions.

Logistic Regression is suitable for
small datasets with linearly separable features,
with a moderate number of informative features
low redundancy.

Decision Trees are effective for large datasets with non-linear decision boundaries and a mix of relevant and irrelevant features.when there are few informative features among many irrelevant ones, or when there are complex interactions between features.

